numpy/core/fromnumeric.py

```diff
@@ -3832,10 +3833,18 @@ def product(*args, **kwargs):
     """
     Return the product of array elements over a given axis.
 
+    .. deprecated:: 1.25.0
+        ``product`` is deprecated as of NumPy 1.25.0, and will be
+        removed in NumPy 2.0. Please use `prod` instead.
+
     See Also
     --------
     prod : equivalent function; see for details.
     """
+    # 2023-03-02, 1.25.0
+    warnings.warn("`product` is deprecated as of NumPy 1.25.0, and will be "
+                  "removed in NumPy 2.0. Please use `prod` instead.",
+                  DeprecationWarning, stacklevel=2)
```

Comment:
I don't want to sweat it, but to note: it is slightly safer to put this into the dispatcher and (conditionally) here.  We do that in some/many/all of such deprecations.
The point being that a downstream user can be affected but not use the NumPy implementation.  (Downstream libs are likely to copy the warning, although if it was on the dispatcher they wouldn't have to.)

Reply:
Hmm, that is true technically. I can't find any example of this in the code base right now. Do you happen to remember an example? If we want to do this structurally, then maybe it's also useful to update the backwards compatibility NEP with a note/example.

Reply:
There is half of an example right now, but the stacking functions used (and still use) the pattern with `overrides.ARRAY_FUNCTION_ENABLED`.

Dunno if we have any other deprecations in flight that should potentially be using the pattern...

They always call the dispatcher (or part of it) to check for sequence in put, that check used to be a warning.

Reply:
Followed up on this in gh-23364